File: Tarea4/procesamientods.py
import sounddevice as sd
import numpy as np
import matplotlib.pyplot as plt


# Menú principal
opcion = 0
while opcion != 5:
    # Menú de opciones
    print('Seleccione una opción:')
    print('1. Grabar')
    print('2. Reproducir')
    print('3. Graficar')
    print('4. Graficar densidad')
    print('5. Salir')

    opcion = int(input('Ingrese su elección: '))

    if opcion == 1:  # Grabación de audio
        try:
            duracion = float(input('Ingrese la duración de la grabación en segundos:  '))
            print('Comenzando la grabación...')
            fs = 44100  # Frecuencia de muestreo
            audio_data = sd.rec(int(duracion * fs), samplerate=fs, channels=1, dtype=np.int16)
            sd.wait()
            print('Grabación finalizada.')
            sd.write('audio.wav', audio_data, fs)
            print('Archivo de audio grabado correctamente')
        except:
            print('Error al grabar el audio')

    elif opcion == 2:  # Reproducción de audio
        try:
            data, fs = sd.read('audio.wav')
            sd.play(data, fs)
            sd.wait()
        except:
            print('Error al reproducir el audio')

    elif opcion == 3:  # Gráfico de audio
        try:
            data, fs = sd.read('audio.wav')
            tiempo = np.linspace(0, len(data) / fs, len(data))
            plt.plot(tiempo, data)
            plt.xlabel('Tiempo (s)')
            plt.ylabel('Amplitud')
            plt.title('Audio')
            plt.show()
        except:
            print('Error al graficar el audio')

    elif opcion == 4:  # Graficando espectro de frecuencia
        try:
            print('Graficando espectro de frecuencia...')
            audio, fs = sd.read('audio.wav')
            N = len(audio)
            f = np.linspace(0, fs / 2, N // 2 + 1)
            ventana = np.hanning(N)
            Sxx, _ = plt.psd(audio, NFFT=N, Fs=fs, window=ventana, noverlap=0)
            plt.xlabel('Frecuencia (Hz)')
            plt.ylabel('Densidad espectral de potencia (dB/Hz)')
            plt.title('Espectro de frecuencia de la señal grabada')
            plt.show()
        except:
            print('Error al graficar el audio.')

    elif opcion == 5:  # Salir
        print('Saliendo del programa...')
    else:
        print('Opción no válida.')

File: Tarea4/ProcesamientoDS.m
%comprueba version
if(exist('OCTAVE_VERSION','builtin')~=0)
%Estamos en octave
  pkg load signal;
end
% Menù principalS
opcion = 0;
while opcion ~= 5
%opcion = input('Seleccione una opciòn:\n 1.Grabar Audio\n 2.Reproducir Audio\n 4. Salir\n');
% Menù de opciones
disp('Seleccione una opcion:')
disp('1. Grabar')
disp('2. Reproducir')
disp('3. Graficar')
disp('4. Grafica densidad')
disp('5. Salir')
opcion = input('Ingrese su elecciòn ');
      switch opcion
      case 1 %Grabaciòn de audio
      try 
         duracion = input('Ingrese la duraciòn de la grabaciòn en segundos');
         disp('Comenzando la grabaciòn...');
         recObj = audiorecorder;
         recordblocking(recObj, duracion);
         disp('Grabaciòn finalizada.');
         data = getaudiodata(recObj.SampleRate);
         audiowrite('audio.wav', data, recObj.SampleRate);
         disp('Archivo de audio grabada correctamente');
      catch
         disp('Error al grabar el audio');
      end
      case 2 %Reproducciòn de audio
      try 
      	[data, fs] = audioread('audio.wav');
      	sound(data, fs);
      catch
      	disp('Error al reproducir el audio');
      end
      case 3 %Gràfico de audio
      try
      	[data, fs] = audioread('audio.wav');
      	tiempo = linspace(0, length(data)/fs, length(data));
      	plot(tiempo, data);
      	xlabel('Tiempo (s)');
      	ylabel('Amplitud');
      	title('Audio');
      catch
      	disp('Error al graficar el audio');
      end
      case 4 %Graficando espectro de frecuencia
      try
      	disp('Graficando espectro de frecuencia...');
      	[audio, Fs] = audioread('audio.wav'); %Lee la señal desde el archivo .wav
      	N = length(audio); %Nùmero de muestras de la señal
      	f = linspace(0, Fs/2, N/2+1); %Vector de frecuencias
      	ventana = hann(N); % Ventanade Hann para reducir el efecto de las discontinuidades al calcular la FFT
      	Sxx = pwelch(audio, ventana, 0, N, Fs); %Densidad espectral de potencia
      	plot(f, 10*log10(Sxx(1:N/2+1))); %Grafica el espectro de frecuencia en dB
      	xlabel('Frecuencia (Hz)');
      	ylabel('Densidad espectral de potencia (dB/Hz)');
      	title('Espectro de frecuencia de la señal grabada');
      catch
      	disp('Error al graficar el audio.');
      end
      case 5 %Salir
      	disp('Saliendo del programa...');
      otherwise
      	disp('Opciòn no vàlida.');
      end
   end	
